qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1996 @@
+        return self._schedule
+
+    @_compile_lazy_circuit_before
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit the builder context and compile the built pulse program."""
+        self._default_alignment_context.__exit__(exc_type, exc_val, exc_tb)
+        self.compile()
+        BUILDER_CONTEXTVAR.reset(self._backend_ctx_token)
+
+    @property
+    def backend(self):
+        """Returns the builder backend if set.
+
+        Returns:
+            Optional[BaseBackend]: The builder's backend.
+        """
+        return self._backend
+
+    @property
+    def block(self) -> Schedule:
+        """Return the active block of this bulder."""
+        return self._block
+
+    @property
+    @_requires_backend
+    def num_qubits(self):
+        """Get the number of qubits in the backend."""
+        return self.backend.configuration().n_qubits
+
+    @property
+    def transpiler_settings(self) -> Mapping:
+        """The builder's transpiler settings."""
+        return self._transpiler_settings
+
+    @transpiler_settings.setter
+    @_compile_lazy_circuit_before
+    def transpiler_settings(self, settings: Mapping):
+        self._compile_lazy_circuit()
+        self._transpiler_settings = settings
+
+    @property
+    def circuit_scheduler_settings(self) -> Mapping:
+        """The builder's circuit to pulse scheduler settings."""
+        return self._circuit_scheduler_settings
+
+    @circuit_scheduler_settings.setter
+    @_compile_lazy_circuit_before
+    def circuit_scheduler_settings(self, settings: Mapping):
+        self._compile_lazy_circuit()
+        self._circuit_scheduler_settings = settings
+
+    @_compile_lazy_circuit_before
+    def compile(self) -> Schedule:
+        """Compile and output the built pulse program."""
+        # Not much happens because we currently compile as we build.
+        # This should be offloaded to a true compilation module
+        # once we define a more sophisticated IR.
+        program = self._schedule.append(self.block, mutate=True)
+        self.set_active_block(Schedule())
+        return program
+
+    @_compile_lazy_circuit_before
+    def set_active_block(self, block: Schedule):
+        """Set the active block for the builder."""
+        assert isinstance(block, Schedule)
+        self._block = block
+
+    @_compile_lazy_circuit_before
+    def append_block(self, block: Schedule):
+        """Add a block to the active schedule block.
+
+        Args:
+            block: Schedule to append to the active schedule block.
+        """
+        self.block.append(block, mutate=True)
+
+    @_compile_lazy_circuit_before
+    def append_instruction(self, instruction: instructions.Instruction):
+        """Add an instruction to the active schedule block.
+
+        Args:
+            instruction: Instruction to append.
+        """
+        self.block.append(instruction, mutate=True)
+
+    def _compile_lazy_circuit(self):
```

Comment:
This is private and all methods that call it have been decorated. The alternative is that instead this method was decorated and all methods that call this would implicitly require a backend? A final way is we could just decorate everything and `_requires_backend` might be called twice. I'm ok on whatever one you prefer.